File: api/generate.py
# -*- coding: utf-8 -*-
"""
创源印记 · Vercel Serverless Function —— 真实 AI 创作接口
路由：POST /api/generate

线上部署：
1. 将本仓库导入 Vercel（New Project → Import Git Repository）
2. 在 Vercel 项目 Settings → Environment Variables 添加：
       DEEPSEEK_API_KEY = sk-xxxx
   （可选）DEEPSEEK_MODEL = deepseek-chat
3. 部署后前端通过同源 /api/generate 调用，无需暴露密钥。

请求体：{"prompt": "创作意图", "creator": "可选"}
返回：  {"ok": true, "content": "...", "model": "...", "usage": {...}}
"""
import json
import os
import time
import urllib.request
import urllib.error
from http.server import BaseHTTPRequestHandler

DEEPSEEK_ENDPOINT = os.environ.get(
    "DEEPSEEK_ENDPOINT", "https://api.deepseek.com/chat/completions"
)
DEEPSEEK_MODEL = os.environ.get("DEEPSEEK_MODEL", "deepseek-chat")

SYSTEM_PROMPT = (
    "你是「创源印记」AI 原生创作确权体系中的创作引擎。用户会给出一个创作意图，"
    "你需要直接产出成品内容，不要解释、不要前后缀寒暄、不要使用 Markdown 代码块。\n"
    "根据创作意图判断体裁：\n"
    "- 若要求诗歌/短诗，输出 4-8 行中文现代诗，意象鲜明；\n"
    "- 若要求海报/视觉/设计方案，输出结构化的视觉方案（主视觉、主标题、副标题、配色、构图要点），每行一条；\n"
    "- 若要求广告/旁白/文案，输出可直接朗读的文案，控制在 80 字以内；\n"
    "- 其他体裁按用户要求产出精炼成品，控制在 200 字以内。\n"
    "只输出作品正文本身。"
)


def _json_response(handler, code, obj):
    raw = json.dumps(obj, ensure_ascii=False).encode("utf-8")
    handler.send_response(code)
    handler.send_header("Content-Type", "application/json; charset=utf-8")
    handler.send_header("Content-Length", str(len(raw)))
    handler.send_header("Cache-Control", "no-store")
    handler.send_header("Access-Control-Allow-Origin", "*")
    handler.send_header("Access-Control-Allow-Headers", "Content-Type")
    handler.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
    handler.end_headers()
    handler.wfile.write(raw)


def _call_deepseek(prompt):
    api_key = os.environ.get("DEEPSEEK_API_KEY", "")
    if not api_key:
        raise RuntimeError("服务端未配置 DEEPSEEK_API_KEY 环境变量")
    payload = {
        "model": DEEPSEEK_MODEL,
        "messages": [
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": prompt},
        ],
        "stream": False,
        "temperature": 0.9,
        "max_tokens": 600,
    }
    data = json.dumps(payload, ensure_ascii=False).encode("utf-8")
    req = urllib.request.Request(
        DEEPSEEK_ENDPOINT,
        data=data,
        headers={
            "Content-Type": "application/json",
            "Authorization": "Bearer " + api_key,
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=60) as resp:
            body = json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        detail = e.read().decode("utf-8", errors="replace")
        raise RuntimeError("DeepSeek API 错误 %s: %s" % (e.code, detail[:400]))
    content = (body.get("choices") or [{}])[0].get("message", {}).get("content", "")
    return {
        "content": content.strip().strip("`").strip(),
        "model": body.get("model", DEEPSEEK_MODEL),
        "usage": body.get("usage", {}),
        "id": body.get("id", ""),
    }


class handler(BaseHTTPRequestHandler):
    def _cors(self):
        self.send_response(204)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.end_headers()

    def do_OPTIONS(self):
        self._cors()

    def do_POST(self):
        try:
            length = int(self.headers.get("Content-Length", "0"))
            raw = self.rfile.read(length) if length else b"{}"
            params = json.loads(raw.decode("utf-8") or "{}")
        except Exception as e:
            return _json_response(self, 400, {"ok": False, "error": "请求体解析失败: %s" % e})

        prompt = (params.get("prompt") or "").strip()
        if not prompt:
            return _json_response(self, 400, {"ok": False, "error": "缺少 prompt 字段"})
        if len(prompt) > 2000:
            return _json_response(self, 400, {"ok": False, "error": "创作意图过长"})

        t0 = time.time()
        try:
            result = _call_deepseek(prompt)
        except Exception as e:
            return _json_response(self, 502, {"ok": False, "error": str(e)})

        result["ok"] = True
        result["latency_ms"] = int((time.time() - t0) * 1000)
        result["server_time"] = int(time.time())
        _json_response(self, 200, result)

File: edge-functions/api/generate.js
/**
 * 创源印记 · EdgeOne Makers 边缘函数 —— 真实 AI 创作接口（后端）
 *
 * 路由（由文件路径自动生成，无需配置）：
 *   POST /api/generate   真实生成（代理 DeepSeek）
 *   GET  /api/generate   健康探针，返回 JSON（用于判断函数是否真的注册成功）
 *   OPTIONS /api/generate  预检
 *
 * 运行环境：EdgeOne Makers Edge Functions（V8 隔离环境，非 Node.js）
 * 可用能力：fetch / Web Crypto / TextEncoder / Response / Request —— 全部为 Web 标准 API
 * 第三方依赖：无（不使用 require、process、fs、npm 包）
 *
 * 需在 EdgeOne Makers 控制台 → 项目设置 → 环境变量 配置：
 *   DEEPSEEK_API_KEY    必需，DeepSeek 密钥（建议设为「加密」类型）
 *   DEEPSEEK_ENDPOINT   可选，默认 https://api.deepseek.com/chat/completions
 *   DEEPSEEK_MODEL      可选，默认 deepseek-chat
 *   ALLOW_ORIGINS       可选，逗号分隔的 Origin 白名单。配置后 POST 仅接受名单内来源，
 *                       用于防止该接口被外部站点/脚本盗用烧 key；不配置则不限制。
 */

const DEFAULT_ENDPOINT = 'https://api.deepseek.com/chat/completions';
const DEFAULT_MODEL = 'deepseek-chat';
const MAX_PROMPT_LEN = 2000;

const SYSTEM_PROMPT =
  '你是「创源印记」AI 原生创作确权体系中的创作引擎。用户会给出一个创作意图，' +
  '你需要直接产出成品内容，不要解释、不要前后缀寒暄、不要使用 Markdown 代码块。\n' +
  '根据创作意图判断体裁：\n' +
  '- 若要求诗歌/短诗，输出 4-8 行中文现代诗，意象鲜明；\n' +
  '- 若要求海报/视觉/设计方案，输出结构化的视觉方案（主视觉、主标题、副标题、配色、构图要点），每行一条；\n' +
  '- 若要求广告/旁白/文案，输出可直接朗读的文案，控制在 80 字以内；\n' +
  '- 其他体裁按用户要求产出精炼成品，控制在 200 字以内。\n' +
  '只输出作品正文本身。';

const CORS_HEADERS = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Headers': 'Content-Type',
  'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
  'Access-Control-Max-Age': '86400',
};

function jsonResponse(obj, status) {
  return new Response(JSON.stringify(obj), {
    status: status || 200,
    headers: Object.assign(
      {
        'Content-Type': 'application/json; charset=utf-8',
        'Cache-Control': 'no-store',
      },
      CORS_HEADERS
    ),
  });
}

function envOf(context) {
  return (context && context.env) || {};
}

/** Origin 白名单校验：未配置 ALLOW_ORIGINS 时放行（向后兼容）。 */
function originAllowed(context) {
  const allow = String(envOf(context).ALLOW_ORIGINS || '').trim();
  if (!allow) return true;
  const list = allow
    .split(',')
    .map(function (s) {
      return s.trim().replace(/\/+$/, '');
    })
    .filter(Boolean);
  const origin = (context.request.headers.get('Origin') || '').replace(/\/+$/, '');
  return origin !== '' && list.indexOf(origin) >= 0;
}

function stripFences(s) {
  return String(s || '')
    .trim()
    .replace(/^```[a-zA-Z]*\s*/, '')
    .replace(/```$/, '')
    .trim();
}

async function callDeepSeek(prompt, context) {
  const env = envOf(context);
  const apiKey = env.DEEPSEEK_API_KEY || '';
  if (!apiKey) throw new Error('服务端未配置 DEEPSEEK_API_KEY 环境变量');
  const endpoint = env.DEEPSEEK_ENDPOINT || DEFAULT_ENDPOINT;
  const model = env.DEEPSEEK_MODEL || DEFAULT_MODEL;

  const resp = await fetch(endpoint, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: 'Bearer ' + apiKey,
    },
    body: JSON.stringify({
      model: model,
      messages: [
        { role: 'system', content: SYSTEM_PROMPT },
        { role: 'user', content: prompt },
      ],
      stream: false,
      temperature: 0.9,
      max_tokens: 600,
    }),
  });

  const raw = await resp.text();
  if (!resp.ok) {
    throw new Error('DeepSeek API 错误 ' + resp.status + ': ' + raw.slice(0, 400));
  }
  let body;
  try {
    body = JSON.parse(raw);
  } catch (e) {
    throw new Error('DeepSeek 返回非 JSON 内容');
  }
  const choice = (body.choices || [{}])[0] || {};
  const content = (choice.message || {}).content || '';
  return {
    content: stripFences(content),
    model: body.model || model,
    usage: body.usage || {},
    id: body.id || '',
  };
}

async function handler(context) {
  const method = (context.request.method || 'GET').toUpperCase();

  if (method === 'OPTIONS') {
    return new Response(null, { status: 204, headers: CORS_HEADERS });
  }

  // 健康探针：函数若未注册成功，该路径会回落到静态资源并返回 HTML，据此可判定注册状态
  if (method === 'GET') {
    return jsonResponse({
      ok: true,
      service: 'chuangyuan-yinji / api/generate',
      runtime: 'edgeone-makers-edge-function',
      key_configured: !!envOf(context).DEEPSEEK_API_KEY,
      origin_restricted: !!String(envOf(context).ALLOW_ORIGINS || '').trim(),
    });
  }

  if (method !== 'POST') {
    return jsonResponse({ ok: false, error: '仅支持 POST / GET / OPTIONS' }, 405);
  }

  if (!originAllowed(context)) {
    return jsonResponse({ ok: false, error: '来源不在 ALLOW_ORIGINS 白名单内' }, 403);
  }

  let params;
  try {
    params = await context.request.json();
  } catch (e) {
    return jsonResponse({ ok: false, error: '请求体解析失败，需为 JSON' }, 400);
  }

  const prompt = String((params && params.prompt) || '').trim();
  if (!prompt) return jsonResponse({ ok: false, error: '缺少 prompt 字段' }, 400);
  if (prompt.length > MAX_PROMPT_LEN) {
    return jsonResponse({ ok: false, error: '创作意图过长（>' + MAX_PROMPT_LEN + ' 字）' }, 400);
  }

  const t0 = Date.now();
  try {
    const result = await callDeepSeek(prompt, context);
    result.ok = true;
    result.latency_ms = Date.now() - t0;
    result.server_time = Math.floor(Date.now() / 1000);
    return jsonResponse(result, 200);
  } catch (e) {
    return jsonResponse({ ok: false, error: String((e && e.message) || e) }, 502);
  }
}

export default handler;
export const onRequest = handler;
